arbitrator/prover/src/lib.rs

```diff
@@ -323,16 +330,34 @@ pub unsafe extern "C" fn arbitrator_set_global_state(mach: *mut Machine, gs: Glo
     (*mach).set_global_state(gs);
 }
 
+#[repr(C)]
+pub struct ResolvedPreimage {
+    pub ptr: *mut u8,
+    pub len: isize, // negative if not found
+}
+
 #[no_mangle]
-pub unsafe extern "C" fn arbitrator_add_preimage(
+pub unsafe extern "C" fn arbitrator_set_preimage_resolver(
```

Comment:
If passing context per machine, user shouldn't also need to pass resolver function per machine (function pointer could be set in machine creation). The function is always constant anyway..

Reply:
We don't have a good place to put the context-less preimage resolver function in the machine before it's ready to resolve preimages. We could add such a field, but I think it's better to keep it open by setting the function here.

Reply:
The ideal function might be statically linking against the Go preimage resolver, but that would complicate producing a standalone prover executable for testing.